### DISCUSSION TITLE
4주차 Discussion

### DISCUSSION THREAD

**Discussion by chloe-codes1** [slippStudy/java-network-study#10] [General]
## 📢 책을 읽으면서 궁금한 점들을 자유롭게 적고 대답해 주세요

4주차 진행 방식

> 진행자: 동훈님
> 

- 7 ~ 8장 읽기
- Discussion 에 질문 or 인상 깊은 점 2개 적어오기

**Comment by jinwookh** [comment]
1.  8장 ChatClinet, ChatServer 구현이 흥미로웠다.
기회가 된다면 다음 두 가지를 하면 좋을 것 같다.
1)) 여러 방이 있는 채팅방 만들기
2)) 요청당 1 쓰레드가 아닌, 멀티플렉싱 다중서버 기반으로 만들어보기
참고자료: https://engineering.linecorp.com/ko/blog/do-not-block-the-event-loop-part2

2. 예제 8-9 WinChatClient에서 
display.append(line + "\n"); 코드가 실행될 때 actionPerfomed 메서드가 호출되는 것일까?
actionPerfomed의 호출하는 시점이 궁금하다.

**Comment by LenKIM** [comment]
InputStream 과 OutputStream 의 실제 활용예시를 TCP 를 통해 이해할 수 있다는 사실이 놀랍다.
byte 단위로 객체를 생각할 수 있는 기반이 만들어진것 같다.

질문거리는 좀더 고민해보기.

**Reply by LenKIM** [reply]
메세지를 보낼 때, 바이트 형태로 떨어질 텐데, 이때 Header가 있나? 있어서 checksum 을 하나?

**Reply by LenKIM** [reply]
실제로 HTTP 통신이 예제처럼 동작하는게 맞을까? 우리가 모르는 개념이 추가되어있나?

**Comment by chloe-codes1** [comment]
### 궁금한 점

1. Java에서 TCP 서버를 구현할 때 어떤 thread 사용 패턴을 사용해봤는지?
  e.g. Thread per client, Thread Pool, Non-blocking I/O, Event-Driven
2. 8장 마지막에서 객체 직렬화는 편리하지만 무거우므로 자원을 적게 사용하고자 할 때에는 권장하지 않는다는 말이 있는데, 편리함 vs 최적화의 trade-off에 대해 어떤 생각을 가지고 있는지 궁금합니다!

**Comment by kangbada0728** [comment]
궁금한점
1. 쓰레드를 만들때 ConcurrentHashMap 을 쓰면 안되는가
2. 객체 직렬화는 사용 안하는게 좋지 않을까?
